hw/misc/hermes.c

```diff
@@ -399,18 +497,61 @@ static uint64_t hermes_bar0_read(void *opaque, hwaddr addr, unsigned size)
             val &= 0xFFFF;
             break;
         }
+    } else if (addr >= 0x1000 &&
+               addr + size <= 0x1000 +
+               sizeof(struct hermes_bar0_command) * UBPF_ENGINES) {
+        ptr = (uint32_t *) &((uint8_t *) &bar0_init.commands)[addr];
+        val = *ptr;
+    } else if (addr >= 0x2000 && addr + size <= 0x2000 + UBPF_ENGINES) {
+        ptr = (uint32_t *) &((uint8_t *) &bar0_init.exec)[addr];
+        val = *ptr;
     } else {
         hermes_bar_warn_invalid(0, addr);
     }
 
     return val;
 }
 
-/* BAR 0 is read-only */
+/* BAR 0 is partly read-only */
 static void hermes_bar0_write(void *opaque, hwaddr addr, uint64_t val,
                 unsigned size)
 {
-    hermes_bar_warn_read_only(0, addr);
+    if (addr + size > sizeof(struct hermes_bar0)) {
+        hermes_bar_warn_invalid(0, addr);
+        return;
+    }
+
+    if (addr >= 0x1000 && addr < 0x2000) {
+        int engine = (addr - 0x1000) / 48;
+        int offset = (addr - 0x1000) % 48;
+
+        if (offset != 0 || engine > UBPF_ENGINES - 1) {
+            hermes_bar_warn_invalid(0, addr);
+            return;
+        }
+
+        memcpy((void *) &bar0_init + addr, &val, size);
+    }
+
+    else if (addr >= 0x2000) {
+        int engine = addr - 0x2000;
+        HermesState *hermes = opaque;
+
+        if (engine > UBPF_ENGINES - 1) {
+            hermes_bar_warn_invalid(0, addr);
+            return;
+        }
+
+        if (val == 1) {
+            bar0_init.exec[engine] = 1;
+            hermes_exec(hermes, engine);
```

Comment:
The eBPF program should be executed into a different thread, otherwise we block the main qemu. Take a look at the XDMA code for an example. Good starting points are `hermes_bar2_init_dir_reg` and `hermes_dma_thread`.